### ISSUE TITLE
Add a cron test to make sure the dependency constraints for build_runner example are up to date

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/site-webdev#1570]
I noticed that on this page https://webdev-dartlang-org-dev.firebaseapp.com/tools/build_runner we give people specific versions to depend on for `build_runner`, `build_web_compilers`, and `build_test`. It is important that these versions are kept up to date so that they can at least get the latest version of all the packages.

I propose a test which runs in a cron job which hits the pub site for each package, and uses `pub_semver` to make sure the latest version is compatible with the current recommended constraints.

**Comment by jakemac53** [comment]
cc @kevmoo @natebosch

**Comment by jakemac53** [comment]
Note that there is another, possibly even more desirable option here, which is to just tell people to use `any` for all these version constraints.

This should generally be very safe to do, the only downside i can see is that if the `build.yaml` config for one of these packages changes you could get broken. Users generally will not depend on the imperative apis provided by these packages though, and the cli is unlikely to change in breaking ways.

**Comment by chalin** [comment]
As you know `^x.y.z` will pick up the latest package. I do update the minimal package version in all example pubspec.yaml files, when I become aware that it is necessary.

FYI, the package versions shown in [that excerpt](https://webdev-dartlang-org-dev.firebaseapp.com/tools/build_runner), are more up-to-date than, say, the angular_router dependencies :). https://github.com/dart-lang/angular/blob/7765221adcca20d5720e97c889f374956fca2f77/angular_router/pubspec.yaml#L19-L21

**Comment by jakemac53** [comment]
> when I become aware that it is necessary.

This is the problem though, how do we make sure the change lands as soon as possible? If we wait for bug reports or complaints then that means a bunch of other users have probably already hit the issue.


> FYI, the package versions shown in that excerpt, are more up-to-date than, say, the angular_router dependencies :).

Sure, but its the same stable version which is the important part ;).

**Comment by jakemac53** [comment]
My concern here, is that in my experience these types of things tend to bitrot because nobody complains loudly about them. I would prefer that we either leave them all as `any` or put in a test to ensure they don't bitrot.

**Comment by chalin** [comment]
TL;DR Maybe you weren't aware but code excerpts shown on the webdev site can't bitrot because they are automatically excerpt from sources that are tested on Travis for every PR. Was that your concern?

---

While we don't have a cron job in place, on every Travis build we already run a **package version check** against our list of "most recent angular related package" versions stored here: https://github.com/dart-lang/site-webdev/blob/master/src/_data/pubspec.yaml. We use `src/_data/pubspec.yaml` to determine which API docs to generate, for example.

Running the package version check now results in:

```nocode
> gulp ng-pkg-pub-upgrade
...
[05:52:17] Using gulpfile ~/git/dart/site-webdev/gulpfile.js
[05:52:17] Starting 'ng-pkg-pub-upgrade'...
[05:52:17] > pub upgrade # cwd: src/_data
[05:52:25] Resolving dependencies...
Overriding the upper bound Dart SDK constraint to <=2.0.0-dev.53.0 for the following packages:
...
Angular package updates available:
  build_web_compilers 0.3.8 (0.4.0 available).
Aborting. Update pubspec(s) before proceeding.
```

So now Travis builds will fail until I update the corresponding package versions in `src/_data/pubspec.yaml`. Unless building and running the example apps fails, _or_ I'm told to use a specific package version (for whatever reason), I don't update the package versions in the example app pubspec.yaml files. This is because pubspec package versions are intended to specify _necessary_ ("minimal") package versions / constraints. From the [pub version constraints page](https://www.dartlang.org/tools/pub/dependencies):

> ... it is important to specify version constraints. That lets people using your package know which versions of its dependencies they can rely on to be compatible with your library. Your goal is to allow a range of versions as wide as possible to give your users flexibility. But it should be narrow enough to exclude versions that you know don’t work or haven’t been tested.

---

That being said, if you still feel we should put "any", and if @kevmoo and @kwalrath agree, I'll make the change (though it might not happen until after I/O).

**Comment by chalin** [comment]
Oh, another alternative is for the pubspec.yaml excerpt shown in https://webdev-dartlang-org-dev.firebaseapp.com/tools/build_runner, to be from https://github.com/dart-lang/site-webdev/blob/master/src/_data/pubspec.yaml. That way, they will always be the most recent versions. Would that work?

**Comment by jakemac53** [comment]
Ah, the above looks like it covers what I was asking for, assuming it is checking that not just angular but all dependencies are able to fetch the latest. Ideally that check would run via a cron job but I think we have enough churn for this site that it would get captured relatively quickly? The cron jobs are pretty easy to set up though.

> Oh, another alternative is for the pubspec.yaml excerpt shown in https://webdev-dartlang-org-dev.firebaseapp.com/tools/build_runner, to be from https://github.com/dart-lang/site-webdev/blob/master/src/_data/pubspec.yaml. That way, they will always be the most recent versions. Would that work?

That seems like it would make sense, if that is the only place that is getting tested like this.

**Comment by jakemac53** [comment]
Oh on further review I see in your example it exactly caught the new build_web_compilers release so yes that seems like exactly what I was asking for here.

**Comment by chalin** [comment]
> I think we have enough churn for this site that it would get captured relatively quickly?

Yes, but I'll look into setting up a cron job once things quite down a bit :).

> ... yes that seems like exactly what I was asking for here.

Great!

**Comment by kwalrath** [comment]
I'm a bit concerned that we won't get that cron job set up. @chalin, are you OK with using `any` for the build_* packages?

**Comment by chalin** [comment]
I can look into it after the Flutter site launch, ok?

> are you OK with using any for the build_* packages?

Sure, if the consensus is to use `any`. (I'd be curious to know what @kevmoo has to say about this.)

**Comment by kevmoo** [comment]
As long as we don't tell users to use `any` – it's fine.

**Comment by chalin** [comment]
Folks often replicate what they see, which is why I'd rather not use `any`, but I'm certainly ok going with the majority here.

This discussion of the use of `any` seems tangential: originally this issue was about setting up a cron job to check package versions. We already have scripts that do the required version checks. All I need to so is setup the cron job itself. I can certainly do so within the next two weeks. Is that timeframe ok? (This hasn't seemed like a high priority precisely because we _are_ alreadying doing the required checks, but we do so on each commit/PR -- which, admittedly, might not be as often as we'd like.)
